### PR TITLE
vxlan offload:

### PR DESCRIPTION
**Pull request by rbaryanai**
put all rte_flow offload in a single module.
moved all existing dpdk offload code to the new module.
created offload ports for vport.

### CODE REVIEW

File: lib/netdev-rte-offloads.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2016 Mellanox Technologies, Ltd.
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.h]: 2019 :)

Comment (asafpenso) [inline lib/netdev-rte-offloads.h]: 2019

File: lib/netdev-rte-offloads.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * matching the offload in_port
+ * struct netdev   - point to the netdev
+ */
+int netdev_rte_offload_add_port(odp_port_t dp_port, struct netdev * netdev);
+
+int netdev_rte_offload_del_port(odp_port_t dp_port);
+
+
+int
+netdev_dpdk_flow_put(struct netdev *netdev, struct match *match,
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.h]: Looks like it should be with the previous line

File: lib/netdev-offload-api.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+#include "netdev-rte-offloads.h"
+
+#define NETDEV_OFFLOAD_ADD_PORT(X,Y) netdev_rte_offload_add_port(X,Y)
+#define NETDEV_OFFLOAD_DEL_PORT(X) netdev_rte_offload_del_port(X)
+
+#define OFFLOAD_RESERVED_MARK (64)
+
+#else
+
```
Comment (asafpenso) [inline lib/netdev-offload-api.h:14]: Dont you need to define NETDEV_OFFLOAD_ADD_PORT/DEL_PORT here as well as "empty" code?

File: lib/dpif-netdev.c [hunk 4]
```diff
@@ -1829,7 +1831,9 @@ dpif_netdev_port_add(struct dpif *dpif, struct netdev *netdev,
     if (!error) {
         *port_nop = port_no;
         error = do_add_port(dp, dpif_port, netdev_get_type(netdev), port_no);
+        NETDEV_OFFLOAD_ADD_PORT(port_no, netdev);
```
Comment (asafpenso) [inline lib/dpif-netdev.c:1834]: what about a check that there's no error after do_add_port?

File: lib/dpif-netdev.c [hunk 5]
```diff
@@ -2058,7 +2063,7 @@ dp_netdev_pmd_find_dpcls(struct dp_netdev_pmd_thread *pmd,
     return cls;
 }
 
-#define MAX_FLOW_MARK       (UINT32_MAX - 1)
+#define MAX_FLOW_MARK       (1 << 24)
```
Comment (asafpenso) [inline lib/dpif-netdev.c:2066]: 1<<24 is 16M, would it be enough?

File: lib/netdev-offload-api.h [hunk 6]
```diff
@@ -0,0 +1,21 @@
+#ifndef _NETDEV_OFFLOAD_API_H_
+#define _NETDEV_OFFLOAD_API_H_
+
+#ifdef DPDK_NETDEV
+
+#include "netdev-rte-offloads.h"
+
+#define NETDEV_OFFLOAD_ADD_PORT(X,Y) netdev_rte_offload_add_port(X,Y)
```
Comment (asafpenso) [inline lib/netdev-offload-api.h:8]: What is the reason of the design to be with Defines and not function pointers?

File: lib/netdev-dpdk.h [hunk 7]
```diff
@@ -43,4 +58,8 @@ free_dpdk_buf(struct dp_packet *buf OVS_UNUSED)
 
 #endif
 
+bool netdev_dpdk_is_dpdk_class(const struct netdev_class *class);
+
+int  netdev_dpdk_get_port_id(const struct netdev * netdev);
```
Comment (asafpenso) [inline lib/netdev-dpdk.h:58]: Omit the space between * and netdev.
Doesn't the check warn about it?

File: lib/netdev-dpdk.h [hunk 8]
```diff
@@ -22,12 +22,27 @@
...
+            NULL,                                                  \
+            NULL,                                                  \
+            NULL,                                                  \
+            NULL,                                                  \
+            netdev_vport_flow_put,                                 \
+            NULL,                                                  \
+            netdev_vport_flow_del,                                 \
+            netdev_vport_init_flow_api
+
+
```
Comment (asafpenso) [inline lib/netdev-dpdk.h]: Please remove all the extra lines

File: lib/netdev-vport.c [hunk 9]
```diff
@@ -970,14 +974,20 @@ netdev_vport_get_ifindex(const struct netdev *netdev_)
 
     return linux_get_ifindex(name);
 }
-
 #define NETDEV_VPORT_GET_IFINDEX netdev_vport_get_ifindex
+#ifdef DPDK_NETDEV
+#define NETDEV_FLOW_OFFLOAD_API DPDK_VPORT_FLOW_OFFLOAD_API 
+#else
+
```
Comment (asafpenso) [inline lib/netdev-vport.c:981]: Can be omitted

File: lib/netdev-rte-offloads.c [hunk 10]
```diff
@@ -0,0 +1,1471 @@
...
+
+#define RTE_FLOW_MAX_TABLES (31)
+#define HW_OFFLOAD_MAX_PHY (128)
+#define INVALID_ODP_PORT (-1)
+
+struct rte_flow;
+struct netdev_rte_port ;
+
+static int netdev_dpdk_validate_flow(const struct match *match);
+static struct rte_flow *
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: Add empty line :)

File: lib/netdev-rte-offloads.c [hunk 11]
```diff
@@ -0,0 +1,1471 @@
...
+    ufid_hw_offload = ufid_hw_offload_remove(ufid, &rte_port->ufid_to_rte);
+    if (ufid_hw_offload) {
+        netdev_rte_port_ufid_hw_offload_free(ufid_hw_offload );
+    }
+
+    return 0;
+}
+int netdev_vport_init_flow_api(struct netdev * netdev OVS_UNUSED)
+{
+            return 0;
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: too many spaces

**Review by asafpenso — COMMENTED** [review]
Please see my comments

File: lib/netdev-rte-offloads.h (hunk 2 above)
Comment (rbaryanai) [reply lib/netdev-rte-offloads.h]: done

File: lib/dpif-netdev.c (hunk 5 above)
Comment (rbaryanai) [reply lib/dpif-netdev.c:2066]: HW max mark is 24 bits.

File: lib/netdev-offload-api.h (hunk 6 above)
Comment (rbaryanai) [reply lib/netdev-offload-api.h:8]: The add port itself is a function pointer.

File: lib/netdev-vport.c (hunk 9 above)
Comment (rbaryanai) [reply lib/netdev-vport.c:981]: it will not compile if user won't use DPDK

Comment (asafpenso) [reply lib/netdev-vport.c:981]: Just the empty line :)

**Pull request closed without merging**